.github/workflows/main.yml

```diff
@@ -0,0 +1,82 @@
+name: taisei-build-test
+
+on:
+  push:
+    branches:
+      - master
+    paths:
+      - "**.c"
+      - "**.h"
+      - "**.yml"
+      - "**.yaml"
+  pull_request:
+    paths:
+      - "**.c"
+      - "**.h"
+      - "**.yml"
+      - "**.yaml"
+
+jobs:
+  linux-test-build:
+    runs-on: ubuntu-20.04 # so we're not using ancient tools
+    steps:
+    - name: Checkout Code
+      uses: actions/checkout@v2
+
+    - name: Checkout Submodules
+      run: git submodule update --init --recursive
+
+    - name: Install Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
+
+    # needed because apt version is too old
+    - name: Install Build Tools (pip)
+      run: pip install meson ninja
+
+    - name: Install Build Tools (apt)
+      run: sudo apt install cmake build-essential libsdl2-dev libsdl2-mixer-dev libogg-dev libopusfile-dev libpng-dev libzip-dev libx11-dev libwayland-dev imagemagick python3-docutils libwebp-dev
+
+    - name: Configure (Meson)
+      run: meson setup build/ -Dinstall_relative=true -Dbuildtype=debug -Db_ndebug=false -Db_lto=false -Dstrip=false -Ddeveloper=true
```

Comment:
Maybe it makes more sense to do a release build here, since that is the least tested configuration during development. We also disable precompiled headers on our OBS builds for that reason, to make sure nothing breaks without them. Not sure why the `install_relative`, but it probably doesn't matter much if we aren't publishing the build artifacts.

Reply:
Maybe also add `--wrap-mode=nofallback` to ensure system dependencies are used.

Reply:
I'd also do `-Dwerror=true -Ddeprecation_warnings=ignore`

Reply:
`-Dwerror=true` will make many, many parts of the code fail due to warnings, including submodules like `basisu`. Is that something we really want for a CI pipeline?

Ex:

```
FAILED: src/25a6634@@taisei@exe/renderer_glcommon_opengl.c.o 
cc -Isrc/25a6634@@taisei@exe -Isrc -I../src -I../subprojects/basis_universal/transcoder -Isubprojects/cglm/include -I../subprojects/cglm/include -I../subprojects/koishi/include -I../subprojects/glad/include -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/SDL2 -I/usr/include/opus -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Werror -std=gnu11 -O3 -D_REENTRANT -DBASISD_IS_BIG_ENDIAN=0 -DBASISD_USE_UNALIGNED_WORD_READS=0 -DBASISD_SUPPORT_ASTC=1 -DBASISD_SUPPORT_ATC=1 -DBASISD_SUPPORT_BC7=1 -DBASISD_SUPPORT_BC7_MODE5=1 -DBASISD_SUPPORT_DXT1=1 -DBASISD_SUPPORT_DXT5A=1 -DBASISD_SUPPORT_ETC2_EAC_A8=1 -DBASISD_SUPPORT_ETC2_EAC_RG11=1 -DBASISD_SUPPORT_FXT1=0 -DBASISD_SUPPORT_PVRTC1=0 -DBASISD_SUPPORT_PVRTC2=1 -DBASISD_SUPPORT_UASTC=1 -Wall -Wpedantic -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Wabsolute-value -Wcast-align -Wcast-align=strict -Wcast-function-type -Wclobbered -Wduplicated-branches -Wduplicated-cond -Wformat-security -Wignored-qualifiers -Winit-self -Wlogical-op -Wmissing-prototypes -Wno-implicit-fallthrough -Wno-long-long -Wno-missing-braces -Wnull-dereference -Wparentheses -Wshadow=compatible-local -Wstrict-prototypes -Wtype-limits -Wunreachable-code -fexcess-precision=standard -fmerge-all-constants -fno-math-errno -fno-signaling-nans -fno-trapping-math -Wno-deprecated-declarations -msse -mfpmath=sse -fpch-preprocess -include taisei_pch.h -MD -MQ 'src/25a6634@@taisei@exe/renderer_glcommon_opengl.c.o' -MF 'src/25a6634@@taisei@exe/renderer_glcommon_opengl.c.o.d' -o 'src/25a6634@@taisei@exe/renderer_glcommon_opengl.c.o' -c ../src/renderer/glcommon/opengl.c
In file included from ../src/hashtable.inc.h:8,
                 from ../src/hashtable_incproxy.inc.h:10,
                 from ../src/hashtable_predefs.inc.h:31,
                 from ../src/hashtable.h:77,
                 from ../src/util/kvparser.h:16,
                 from ../src/util.h:22,
                 from ../src/global.h:17,
                 from ../src/pch/taisei_pch.h:14:
In function ‘glcommon_ext_flag’,
    inlined from ‘glcommon_check_extension’ at ../src/renderer/glcommon/opengl.c:81:20:
../src/log.h:132:34: error: ‘%s’ directive argument is null [-Werror=format-overflow=]
  132 |  #define _do_log(func, lvl, ...) (func)(lvl, __func__, "<unknown>", 0, __VA_ARGS__)
      |                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/log.h:142:24: note: in expansion of macro ‘_do_log’
  142 | #define log_fatal(...) _do_log(_taisei_log_fatal, LOG_FATAL, __VA_ARGS__)
      |                        ^~~~~~~
../src/renderer/glcommon/opengl.c:55:3: note: in expansion of macro ‘log_fatal’
   55 |   log_fatal("Bad extension string: %s", ext);
      |   ^~~~~~~~~
../src/renderer/glcommon/opengl.c: In function ‘glcommon_check_extension’:
../src/renderer/glcommon/opengl.c:55:36: note: format string is defined here
   55 |   log_fatal("Bad extension string: %s", ext);
      |                                    ^~
cc1: all warnings being treated as errors
[39/284] Compiling C object 'src/25a6634@@taisei@exe/boss.c.o'.
[40/284] Compiling C++ object 'subprojects/basis_universal/1d967b7@@basisu_transcoder@sta/transcoder_basisu_transcoder.cpp.o'.
FAILED: subprojects/basis_universal/1d967b7@@basisu_transcoder@sta/transcoder_basisu_transcoder.cpp.o 
c++ -Isubprojects/basis_universal/1d967b7@@basisu_transcoder@sta -Isubprojects/basis_universal -I../subprojects/basis_universal -fvisibility=hidden -fdiagnostics-color=always -DNDEBUG -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wnon-virtual-dtor -Werror -std=c++14 -fno-exceptions -fno-rtti -O3 -fPIC -Wall -Wno-class-memaccess -Wno-misleading-indentation -Wno-sign-compare -Wno-unused-but-set-variable -Wno-unused-variable -fno-exceptions -fno-rtti -fno-strict-aliasing -DBASISD_IS_BIG_ENDIAN=0 -DBASISD_USE_UNALIGNED_WORD_READS=0 -DBASISD_SUPPORT_ASTC=1 -DBASISD_SUPPORT_ATC=1 -DBASISD_SUPPORT_BC7=1 -DBASISD_SUPPORT_BC7_MODE5=1 -DBASISD_SUPPORT_DXT1=1 -DBASISD_SUPPORT_DXT5A=1 -DBASISD_SUPPORT_ETC2_EAC_A8=1 -DBASISD_SUPPORT_ETC2_EAC_RG11=1 -DBASISD_SUPPORT_FXT1=0 -DBASISD_SUPPORT_PVRTC1=0 -DBASISD_SUPPORT_PVRTC2=1 -DBASISD_SUPPORT_UASTC=1 -MD -MQ 'subprojects/basis_universal/1d967b7@@basisu_transcoder@sta/transcoder_basisu_transcoder.cpp.o' -MF 'subprojects/basis_universal/1d967b7@@basisu_transcoder@sta/transcoder_basisu_transcoder.cpp.o.d' -o 'subprojects/basis_universal/1d967b7@@basisu_transcoder@sta/transcoder_basisu_transcoder.cpp.o' -c ../subprojects/basis_universal/transcoder/basisu_transcoder.cpp
../subprojects/basis_universal/transcoder/basisu_transcoder.cpp: In function ‘bool basist::transcode_uastc_to_bc7(const basist::unpacked_uastc_block&, basist::bc7_optimization_results&)’:
../subprojects/basis_universal/transcoder/basisu_transcoder.cpp:12710:26: error: ‘*((void*)& best_pbits +4)’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
12710 |    dst_blk.m_pbits[0][1] = best_pbits[1];
      |    ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
../subprojects/basis_universal/transcoder/basisu_transcoder.cpp:12709:26: error: ‘best_pbits’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
12709 |    dst_blk.m_pbits[0][0] = best_pbits[0];
      |    ~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
ninja: build stopped: subcommand failed.
```

Reply:
Meson has gained an ability to set `werror` per-subproject, though I'm not sure since when. If configuring initially with `werror=true` defaults it to true for subprojects as well, then we should be able to set it just for the main project as a separate step, e.g.
```
$ meson setup -Dfoo=bar build/
$ meson configure build/ -Dwerror=true
$ meson compile -C build/
```

Alternatively, you should be able to disable werror for specific subprojects manually:
```
$ meson setup -Dfoo=bar -Dwerror=true -Dbasis_universal:werror=false build/
$ meson compile -C build/
```

If neither of this works, then meson is too old. Ideally we should be testing against our minimum supported version, but I'd like to bump that requirement to something more reasonable anyway.

As for warnings that occur in the main project, they must be just fixed (or silenced, in cases where the compiler is being dumb).

Reply:
Got it. Running it on my own private fork, it seems that `src/coroutines.c` has some hard warnings in it:

```
../src/coroutine.c:150:35: error: statement with no effect [-Werror=unused-value]
 #define STAT_VAL_SET(name, value) (0)
```

macOS also fails with a different warning.

You can see here: https://github.com/StarWitch/taisei/pull/1/checks?check_run_id=1574067166

If we want, we could have two runners: one that's more permissive for most things ("this actually compiles"), and one that hard-fails on these sorts of warning-errors ("code health"). The runners don't take too long so I don't think we're at a risk of running out of free CI minutes.

Reply:
All fixed in #273.